### PR TITLE
Use Espree instead of esprima to support newer ES specs

### PR DESCRIPTION
**Pull request by hummlas**
All tests still pass and I verified that it works by npm linking this branch to a local version of cardinal, which I then linked to a project where I successfully highlighted code that use the stage 2 proposal for object spread syntax.

Fixes thlorenz/cardinal#16

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,6 +28,6 @@
     "cardinal": "~0.4.4"
   },
   "dependencies": {
-    "esprima": "~2.7.0"
+    "esprima-fb": "^15001.1001.0-dev-harmony-fb"
```
Comment (hummlas) [inline package.json]: This is the version number you get when running `npm install esprima-fb --save`

**Comment by ariya** [comment]
Oh, why do we want to go back to the unmaintained fork of Esprima again 😢 
Please read PR #4 for more details.

**Comment by hummlas** [comment]
Didn't know Esprima-fb had been used before in this repo. It was the easiest drop-replacement for Esprima that I could find that supports object-spread operator and JSX syntax.

@ariya Is it possible to get Esprima to support stage 2+ ES features and JSX syntax, or do you know of a Esprima-compatible parser that does?

**Comment by ariya** [comment]
ESTree might work for you.

OTOH, the best way to support various syntax is to have a fallback/tolerant mode: if the input can't be parsed, try to tokenize the input.

**Comment by hummlas** [comment]
I'll give Espree a shot then :)

File: redeyed.js [hunk 2]
```diff
@@ -170,14 +170,57 @@ function bootstrap(esprima, exportFn) {
       .map(function (k) { return all[k]; });
   }
 
+  function deduceSourceType(code, parser) {
```
Comment (hummlas) [inline redeyed.js]: https://github.com/jquery/esprima/issues/1492#issuecomment-215165234

File: test/redeyed-smoke.js [hunk 3]
```diff
@@ -53,6 +53,13 @@ test('esprima', function (t) {
 
   readdirp({ root: esprimadir, fileFilter: '*.js' })
     .on('data', function (entry) {
+      var invalidEspreeFiles = [
+          'lib/visitor-keys.js'
```
Comment (hummlas) [inline test/redeyed-smoke.js]: Line 7: ` * The MIT license. Copyright (c) 2014 Ingvar Stepanyan` causes a false positive for the check `!(~result.indexOf('var '))` due to the name "Ing**var** Stepanyan"

**Comment by hummlas** [comment]
Ok, I've rewritten the PR to use Espree instead of Esprima-fb now.
Should I replace all references to 'esprima' with 'espree' or should I leave it as is?

**Comment by hummlas** [comment]
Hmm.. Some tests seem to fail in travis-ci, but all tests succeed locally..

**Comment by ariya** [comment]
I can't comment much on this due to conflict of interest (I'm the author of Esprima).

Notwithstanding that, I still think the right fix is to fallback to pure tokenization mode. Otherwise, we'll keep switching parser again in the future.

**Comment by hummlas** [comment]
@ariya I'm not even sure what that means, is that something that should be fixed in esprima or in redyed? Could you point me in the general direction of where I should start looking to implement the above mentioned fallback?

**Comment by ariya** [comment]
@hummlas Let me write it up as a new issue.

**Comment by thlorenz** [comment]
Yeah someone else would have to write that tokenizer, I don't have bandwidth for it. Or is _pure tokenization mode_ a feature of esprima feature I'm not aware of @ariya?

IMO if all tests pass + it now supports ES6 features that's a gain, so I'm  for that.

However I'm not happy with the last commit @hummlas, just running less smoke tests is not a good idea. Instead adapt the `.travisci.yml` script to include 

```yml
before_install:
  - npm install -g npm@2
```

[see here](https://github.com/thlorenz/sample-heap-allocations/blob/master/.travis.yml).

Once that's done I'm willing to merge this for now even though that makes @ariya sad.

**Comment by ariya** [comment]
> Or is pure tokenization mode a feature of esprima feature I'm not aware of @ariya?

Yes! It already exists, Esprima can tokenize an entire input stream without parsing it.
This is what I plan to write in more details, because I think this is the proper long-term solution for redeyed.

**Comment by hummlas** [comment]
@thlorenz There were a few files (in the nested node_modules folder) that cause false positives for the occurrence `var` declarations, for instance a comment containing the name 'Ingvar`. 

I tried to just exclude the files that caused the false positives for `var` occurrences, and that solved the problem, at least locally.. When it runs in travis the test times out, probably because there are so many  (1791) files to traverse in the nested node_modules folder (see https://travis-ci.org/thlorenz/redeyed/jobs/152034277).

All that being said it's probably better to go with @ariya's solution in #8 :)

**Comment by thlorenz** [comment]
@hummlas well that's a bug in the parser and thus those tests demonstrate that things break currently if we use it.
We can't start using that parser unless that's fixed, otherwise we'll get these false positives when people use it as well.

> I tried to just exclude the files that caused the false positives

Not a good idea to do that, these files show a bug in the parser. That's exactly why I added the smoke tests to detect those edge cases.

Bottom line is things need to work after the parse change **without** changing any tests.

**Comment by hummlas** [comment]
@thlorenz I don't think it's a bug in the parser. I think it's due to the second part of the assert in the test-case:
```
t.assert(~result.indexOf('+var-') || !(~result.indexOf('var ')
``` 
in [redeyed-smoke.js](https://github.com/thlorenz/redeyed/blob/master/test/redeyed-smoke.js#L60)

The below part should make sure that the 'var' part of all var-statements are replace with '+var-':
```
result = redeyed(code, { Keyword: { 'var': '+:-' } }).code
```
e.g. `var foo;` becomes `+var- foo;`

But the assert checks if the result contains a successfully transformed `var` ('+var-') OR if there were no `var` to transform to begin with, in which case the test is also counted as passing.

The problem is however that `!(~result.indexOf('var ')` will tell us that there were indeed a 'var' that should have been transformed if our code contains something like:
```
 // Created by Ingvar Kamprad
```
even though the 'var ' that was found wasn't part of var-statement, and thus shouldn't have been transformed.


Please correct me if I'm wrong though :)

**Comment by ariya** [comment]
This is now overtaken by event, see PR #9.
